test/integration/otel_test.go

```diff
@@ -0,0 +1,142 @@
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/eggsampler/acme/v3"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+	"go.opentelemetry.io/otel"
+	traceSDK "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/test"
+)
+
+// TraceResponse is just the minimal fields we care about from the Jaeger traces API
+type TraceResponse struct {
+	Data []struct {
+		Spans []struct {
+			OperationName string
+		}
+	}
+}
+
+func getTraceFromJaeger(t *testing.T, traceID trace.TraceID) *TraceResponse {
+	traceURL := "http://jaeger:16686/api/traces/" + traceID.String()
+	resp, err := http.Get(traceURL)
+	test.AssertNotError(t, err, "failed to trace from jaeger: "+traceID.String())
+	test.AssertEquals(t, resp.StatusCode, http.StatusOK)
+
+	body, err := io.ReadAll(resp.Body)
+	test.AssertNotError(t, err, "failed to read trace body")
+
+	var parsed TraceResponse
+	err = json.Unmarshal(body, &parsed)
+	test.AssertNotError(t, err, "failed to decode traces body")
+
+	test.Assert(t, len(parsed.Data) > 0, fmt.Sprintf("expected to get traces for %s: %v (%s)", traceID, parsed, traceURL))
+
+	return &parsed
+}
+
+// assertSpans checks that all the passed in traces contain the list of spans
+func assertSpans(t *testing.T, tr *TraceResponse, spans []string) {
+	for _, foo := range tr.Data {
+		for _, span := range spans {
+			found := false
+			for _, a := range foo.Spans {
+				if span == a.OperationName {
+					found = true
+					break
+				}
+			}
+			test.Assert(t, found, fmt.Sprintf("Didn't find expected span: %s in %v", span, foo))
+		}
+	}
+}
+
+type ContextInjectingRoundTripper struct {
+	ctx context.Context
+	rt  http.RoundTripper
+}
+
+func (c ContextInjectingRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	return c.rt.RoundTrip(request.WithContext(c.ctx))
+}
+
+// TestTraces tests that all the expected spans are present and properly connected
+func TestTraces(t *testing.T) {
+	t.Parallel()
+	if !strings.Contains(os.Getenv("BOULDER_CONFIG_DIR"), "test/config-next") {
+		t.Skip("OpenTelemetry is only configured in config-next")
+	}
+
+	traceID := traceIssuingTestCert(t)
+
+	// Sleep to allow traces to flush
+	// TODO: We could retry with backoff instead, allowing this test to complete faster.
+	// TODO: We could also configure a lower batch delay in CI
+	time.Sleep(1.2 * traceSDK.DefaultScheduleDelay * time.Millisecond)
+
+	// TODO: We really want to assert more structure of the trace here
+	assertSpans(t, getTraceFromJaeger(t, traceID), []string{
+		"/directory",
+		"/acme/new-acct",
+		"sa.StorageAuthority/GetOrderForNames",
+		"/acme/chall-v3/",
+		"nonce.NonceService/Nonce",
+	})
+}
+
+func traceIssuingTestCert(t *testing.T) trace.TraceID {
+	domains := []string{random_domain()}
+
+	// Configure this integration test to trace to jaeger:4317 like Boulder will
+	_, _, oTelShutdown := cmd.StatsAndLogging(
```

Comment:
this caused a data race with the grpc logger.  Probably shouldn't use that function here.